### PR TITLE
feat: support ksp bundling

### PR DESCRIPTION
**Pull request by lucaseck**
### What was the problem/requirement? (What/Why)
There was no pathmapping support with KeyShot. 
### What was the solution? (How)
One way around this is to bundle a scene up as a KeyShot Package(ksp) which creates a zipped archive of the scene file and any external file references. The scene within is modified to have relative paths to all of the referenced files. 

This PR adds a submission mode selector for:
1. Just submit the scene .bip file and all files referenced within will have to be available through network storage or some other means
2. Automatically save the current scene as a ksp to a temp directory, unpack it and then submit the modified scene and assets from the unpacked ksp where all paths are relative
### What is the impact of this change?
KeyShot can be used without having to deal with manually packing/unpacking ksps or setting up network storage
### How was this change tested?
- Ran all unit tests including newly included ones
- Ran submissions using both submission modes and verified the outputs
- Tested canceling the dialog, and the sticky setting of which option was selected
### Was this change documented?
Only in the README
### Is this a breaking change?
No
----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py [hunk 1]
```diff
@@ -357,55 +346,137 @@ def gui_submit(bundle_directory: str) -> Optional[dict[str, Any]]:
...
+    input_filenames = [
+        os.path.join(unpack_dir, file)
+        for file in os.listdir(unpack_dir)
+        if not file.endswith(".bip")
+    ]
+
+    bip_files = glob.glob(os.path.join(unpack_dir, "*.bip"))
+
+    if not bip_files:
+        raise RuntimeError("No .bip files found in the KSP bundle.")
```
Comment (joel-wong-aws) [inline src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:416]: Should we raise an error if there are multiple bip files?

File: test/keyshot_submitter/test_keyshot_submitter.py [hunk 2]
```diff
@@ -242,3 +243,45 @@ def test_settings_apply_submitter_settings():
...
     assert sorted(settings.output_directories) == ["test_directory_2", "test_directory_4"]
     assert sorted(settings.referenced_paths) == ["test_ref_path_2"]
+
+
+def test_save_ksp_bundle():
+    dir = os.path.normpath("/testdir/test")
+    bundle_name = "test_bundle.ksp"
+    expected_bundle_path = os.path.normpath(f"{dir}/{bundle_name}")
+
+    output = submitter._save_ksp_bundle(dir, bundle_name)
```
Comment (joel-wong-aws) [inline test/keyshot_submitter/test_keyshot_submitter.py]: nit: if we're testing this function, should it be a public function (without leading underscore)?

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py [hunk 3]
```diff
@@ -357,55 +346,137 @@ def gui_submit(bundle_directory: str) -> Optional[dict[str, Any]]:
...
+    Returns a dictionary of the selected option values in the format:
+        {'SUBMISSION_MODE_KEY': [1, 'only the scene BIP file']}
+    """
+    dialog_items = [
+        (
+            SUBMISSION_MODE_KEY,
+            lux.DIALOG_ITEM,
+            "Do you want to submit",
+            0,
+            ["the scene BIP file and all external files references", "only the scene BIP file"],
```
Comment (crowecawcaw) [inline src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py]: I think the wording could be tweaked a bit so an artist would know how to choose one setting versus another. Maybe include a description like:
> Choose "the scene BIP file and all external file references" if you're rendering on service-managed fleets or to upload your scene to a customer-managed fleet. Choose "only the scene BIP file" if you're rendering on customer-managed fleets that already have access to the file references, for example on a shared network drive. 

Nit: Also instead of "Do you want to submit", maybe rephrase as "What files do you want to attach with your job?" That way the user selection answers the question instead of form part of the question itself.

File: test/keyshot_submitter/test_keyshot_submitter.py (hunk 2 above)
Comment (lucaseck) [reply test/keyshot_submitter/test_keyshot_submitter.py]: Leftover from before my initial rebase. Removed the leading `_` from all new functions to follow the pattern we've already established in the file.

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py (hunk 1 above)
Comment (lucaseck) [reply src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:416]: I'm not sure that there is ever a case where there can be multiple BIP files, but I've added the error just in case.

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py (hunk 3 above)
Comment (lucaseck) [reply src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py]: There's nothing stopping someone from setting up network storage for service managed fleets as far as I know. I don't think we can make the suggestion of which to use based on if SMF or CMF are being used.

Comment (lucaseck) [reply src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py]: Updated the wording of the prompt itself

**Review by crowecawcaw — DISMISSED** [review]

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py [hunk 4]
```diff
@@ -370,54 +436,62 @@ def main(lux):
...
-    job_template = construct_job_template(filename)
-    asset_references = construct_asset_references(settings)
-    parameter_values = construct_parameter_values(settings)
+    with tempfile.TemporaryDirectory() as bundle_temp_dir:
+        # {'submission_mode': [0, 'the scene BIP file and all external files references']}
+        if not dialog_selections[SUBMISSION_MODE_KEY][0]:
+            temp_scene_file, input_filenames = get_ksp_bundle_files(bundle_temp_dir)
+            settings.auto_detected_input_filenames = input_filenames
+            settings.parameter_values.append({"name": "KeyShotFile", "value": temp_scene_file})
+        else:
```
Comment (crowecawcaw) [inline src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:483]: Do we need to attach the scene file in this branch? Also do we still want to attach external files if KSP bundling isn't selected? (I think not, but just double checking.)

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py [hunk 5]
```diff
@@ -370,54 +436,62 @@ def main(lux):
         else:
             lux.saveFile()
 
-    scene_file = lux.getSceneInfo()["file"]
-    external_files = lux.getExternalFiles()
+    dialog_selections = options_dialog()
+
+    if not dialog_selections:
+        print("Job submission canceled.")
+        return
```
Comment (crowecawcaw) [inline src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:443]: We should eventually raise an exception here so the job status doesn't show Suceess afterwards. Related PR: https://github.com/aws-deadline/deadline-cloud-for-keyshot/pull/102 Whichever PR get's merged first, we should fix the other.

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py (hunk 4 above)
Comment (lucaseck) [reply src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:483]: The scene file parameter is a PATH parameter, just by having it specified it will include it as an input file, no need to actually add it to the asset references directly

Comment (lucaseck) [reply src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:483]: No, if just the .bip file is selected we don't want to attach any of the external files since they should be made available through network storage or some other means.

**Review by crowecawcaw — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -34,6 +34,24 @@ This package provides a KeyShot plugin script that creates jobs for AWS Deadline
...
+There are two submission modes for the KeyShot submitter which a dialog will ask you to select from before opening the submitter UI.
+
+1. Attach `The scene BIP file and all external files references`
+    - The open scene file and all external files referenced within will be included
+    as job attachments. The submitter will export the open scene to a
+    KSP(KeyShot Package) which turns all file paths in the scene into relative
+    paths and creates a flattened directory with all of the external files directly
+    beside the scene file. The KSP is then unzipped, and the new scene file with
+    the relative paths and all external files will be submitted with the job.
+    The temporary directory used to save the KSP will be deleted after each
```
Comment (ryyakobe) [inline README.md:48]: Have we considered not deleting this temp KSP file but saving it with submission timestamp, for example? This way we would create a "snapshot" of the scene for each submission, so it would be easy to jump back to the required revision, if needed. Not sure if there is a real use-case behind it.

Also, if current scene is actually a part of existing KSP file, it seems like we would still create a temp SKP file, update all the paths, create temp directory and delete it. Do we need to have all these steps for existing KSP files?

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py [hunk 7]
```diff
@@ -357,7 +347,83 @@ def gui_submit(bundle_directory: str) -> Optional[dict[str, Any]]:
...
+                  KSP bundle before submission.
+    Returns a dictionary of the selected option values in the format:
+        {'SUBMISSION_MODE_KEY': [1, 'only the scene BIP file']}
+    """
+    dialog_items = [
+        (
+            SUBMISSION_MODE_KEY,
+            lux.DIALOG_ITEM,
+            "What files would you like to attach to the job?",
+            0,
```
Comment (ryyakobe) [inline src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:366]: What does `0` mean here? Is it the default selected "index" in `["The scene BIP file and all external files references", "Only the scene BIP file"]`?

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py [hunk 8]
```diff
@@ -357,7 +347,83 @@ def gui_submit(bundle_directory: str) -> Optional[dict[str, Any]]:
...
+
+def save_ksp_bundle(directory: str, bundle_name: str) -> str:
+    """
+    Saves out the current scene and any file references to a ksp bundle in a
+    directory.
+    Returns the file path where the bundle was saved to.
+    """
+    full_ksp_path = os.path.join(directory, bundle_name)
+
+    if not lux.savePackage(path=full_ksp_path):
```
Comment (ryyakobe) [inline src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:387]: What happens if the file with this name already exists? Would we override it or would we get an error?

What do you think about checking if the file already exists before trying to save it? Does it make sense to add some randomness, like timestamp, to the `temp_deadline_cloud.ksp` filename to avoid name collisions, even if the temp file has not been cleaned up properly? Similar for `unpack` directory.

**Review by ryyakobe — APPROVED** [review]

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py (hunk 7 above)
Comment (lucaseck) [reply src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:366]: 0 is just the index of the options, default to the first option which is the scene file and all external references

File: src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py (hunk 8 above)
Comment (lucaseck) [reply src/deadline/keyshot_submitter/Submit to AWS Deadline Cloud.py:387]: It's being created in a temp directory that is deleted at the end of every submission, there should never be any name collisions.

File: README.md (hunk 6 above)
Comment (lucaseck) [reply README.md:48]: I considered it, but couldn't think of a good enough reason right now to add the extra complexity of managing that. No issue with it being a future improvement if that's a workflow that's requested.

As far as I know if a user tries to open an existing ksp up it has to be extracted by KeyShot somewhere(either the Resources folder or another location) before KeyShot can be used, so everything should be as a bip file during submission
